### PR TITLE
SNOW-1645316: Modularize skip lqb conditions and add no db check

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1645316

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

   i) Move skip large query breakdown into two functions:
   - `_should_skip_optimization_for_session`
   - `_should_skip_optimization_for_root`
 
   ii) Add a condition to skip large query breakdown when no database exists

### CODE REVIEW

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 1]
```diff
@@ -145,6 +145,49 @@ def apply(self) -> List[LogicalPlan]:
 
         return resulting_plans
 
+    def _should_skip_optimization_for_session(
+        self,
+    ) -> Optional[SkipLargeQueryBreakdownCategory]:
+        """Method to check if the optimization should be skipped based on the session state."""
+        if self.session.get_current_database() is None:
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:152]: how about schema? when there no current schema, that won't work also

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 2]
```diff
@@ -123,14 +123,14 @@ def __init__(
         self.complexity_score_upper_bound = complexity_bounds[1]
 
     def apply(self) -> List[LogicalPlan]:
-        if is_active_transaction(self.session):
+        if reason := self._should_skip_optimization_for_session():
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:126]: let's just do 
```
reason = self._should_skip_optimization_for_session()
if reason is not None:
    xxx
```
easier for me to understand

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py [hunk 3]
```diff
@@ -145,6 +145,49 @@ def apply(self) -> List[LogicalPlan]:
 
         return resulting_plans
 
+    def _should_skip_optimization_for_session(
+        self,
+    ) -> Optional[SkipLargeQueryBreakdownCategory]:
+        """Method to check if the optimization should be skipped based on the session state."""
```
Comment (sfc-gh-yzou) [inline src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:151]: add comment about the return value

File: src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py (hunk 1 above)
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/_internal/compiler/large_query_breakdown.py:158]: added a check anyways.

**Review by sfc-gh-jdu — APPROVED** [review]

**Pull request merged**